gym_solo/core/test_obs_observations.py

```diff
@@ -66,5 +66,157 @@ def test_compute(self, mock_vel, mock_base):
+    mock_joint_info.side_effect = joint_info_side_effect
+    dummy_robot_id = 0
+
+    o = obs.MotorEncoder(dummy_robot_id, degrees= degrees)
+
+    if degrees:
+      position_max = 572.9578
+      position_min = -572.9578
+    else:
+      position_max = 10
+      position_min = -10
+
+    lower_bound = np.full(num_joints, position_min)
+    upper_bound = np.full(num_joints, position_max)   
+
+    np.testing.assert_allclose(o.observation_space.low, lower_bound)
+    np.testing.assert_allclose(o.observation_space.high, upper_bound)
+
+
+  @mock.patch('pybullet.getJointInfo', autospec=True)
+  @mock.patch('pybullet.getNumJoints', autospec=True)
+  def test_labels(self, mock_num_joints, mock_joint_info):
+    num_joints = 12
+    dummy_robot_id = 0
+
+    mock_num_joints.return_value = num_joints
+    mock_joint_info.side_effect = self.joint_info
+    
+    ground_truth = ['FL_HFE', 'FL_KFE', 'FL_ANKLE', 'FR_HFE',
+     'FR_KFE', 'FR_ANKLE', 'HL_HFE', 'HL_KFE', 'HL_ANKLE',
+     'HR_HFE', 'HR_KFE', 'HR_ANKLE']
+
+    o = obs.MotorEncoder(dummy_robot_id)
+    self.assertEqual(o.labels(), ground_truth)
+    
+  
+  @parameterized.expand([
+    ("default", False),
+    ("degrees", True)
+  ])
+  @mock.patch('pybullet.getJointState', autospec=True)
+  @mock.patch('pybullet.getNumJoints', autospec=True)
+  def test_compute(self, name, degrees, mock_num_joints, mock_joint_state):
+    num_joints = 12
+    dummy_robot_id = 0
+
+    mock_num_joints.return_value = num_joints
+
+    # This is real case extracted from pybullet
+    mock_joint_state.side_effect = [
+      (1.5301299626083, 7.554435979113249e-11, (0.0, 0.0, 0.0, 0.0, 0.0, 0.0), 0.0), 
+      (-3.0853209964046426, -1.4807635358886225e-11, (0.0, 0.0, 0.0, 0.0, 0.0, 0.0), 0.0), 
+      (0.0, 0.0, (0.0, 0.0, 0.0, 0.0, 0.0, 0.0), 0.0), 
+      (1.530127327627307, 7.721743189663525e-12, (0.0, 0.0, 0.0, 0.0, 0.0, 0.0), 0.0), 
+      (-3.085315909474513, 1.2383787804023483e-11, (0.0, 0.0, 0.0, 0.0, 0.0, 0.0), 0.0), 
+      (0.0, 0.0, (0.0, 0.0, 0.0, 0.0, 0.0, 0.0), 0.0), 
+      (-1.530132288799807, -1.9996261740838904e-11, (0.0, 0.0, 0.0, 0.0, 0.0, 0.0), 0.0), 
+      (3.0853224548246283, -2.1458507061027347e-11, (0.0, 0.0, 0.0, 0.0, 0.0, 0.0), 0.0), 
+      (0.0, 0.0, (0.0, 0.0, 0.0, 0.0, 0.0, 0.0), 0.0), 
+      (1.5301292310246128, -2.8650758804247376e-11, (0.0, 0.0, 0.0, 0.0, 0.0, 0.0), 0.0), 
+      (-3.0853176193095613, 1.4177090742904175e-11, (0.0, 0.0, 0.0, 0.0, 0.0, 0.0), 0.0), 
+      (0.0, 0.0, (0.0, 0.0, 0.0, 0.0, 0.0, 0.0), 0.0)]
+    
+    # Built from the mock_joint_state.side_effect value. Refer to getJointState return from
+    # https://docs.google.com/document/d/10sXEhzFRSnvFcl3XxNGhnD4N2SedqwdAvK3dsihxVUA/edit#heading=h.p3s2oveabizm
+    ground_truth = [1.5301299626083, -3.0853209964046426, 0.0,
+      1.530127327627307, -3.085315909474513, 0.0,
+      -1.530132288799807, 3.0853224548246283, 0.0,
+      1.5301292310246128, -3.0853176193095613, 0.0]
+
+    if degrees:
```

Comment:
Is there a genuine difference? I thought it was a lot cleaner this way as it reduced duplicated code

Reply:
There's a small but important difference. Consider the two examples:

```python
def test_subtest(self):
  # expensive ass operation
  with self.subTest('test1'):
    # asserts
  with self.subTest('test2'):
    # assert that fails

@parameterized.expand([
  ('test1'),('test2')
])
def test_parameterized(self, name):
  # expensive ass operation
  if self.name == 'test1':
    # asserts
  if self.name == 'test2':
    # assert that fails
```

So if the assert fails for subtest2, the program runs through the expensive operation, subtest1, **and with the same context** subtest2. When subtest2 fails, since it was in `subTest` context, `unittest` knows to append that at the end of the name.

`parameterized` on the other hand literally creates multiple copies of the test with different arguments passed in. Thus, for the parameterized test, the expensive operation is actually running **twice**. In this case, it shouldn't matter too much, but it's a good habit to get into. For things when you're actually interacting with pybullet, using `subTest` can significantly speed up the runtime.

Reply:
I'm a dumbass nvm